### ISSUE TITLE
Heap problem when trying to send more than 1 attachment

### ISSUE THREAD

**Issue by mateusz-kusmierz** [mobizt/ESP32-Mail-Client#39]
Does anyone had simmilar problem?
What does it mean and how i could solve it?

Serial print:
```
Number of files attached: 2
Connecting to SMTP server...
Identification...
Authentication...
Sign in...
Sending Email header...
Sending Email body...
Sending attachments...
/SCAN_34.csv
/SCAN_35.csv
CORRUPT HEAP: Bad head at 0x3ffde3d4. Expected 0xabba1234 got 0x7042795a
assertion "head != NULL" failed: file "/home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/heap/multi_heap_poisoning.c", line 214, function: multi_heap_free
abort() was called at PC 0x4013c993 on core 1

```

Sending code:  (private data masked using *****)
```
void sendEmail()                                                     
{

  smtpData.setLogin("smtp.gmail.com", 587, "*******@gmail.com", "*******");
  smtpData.setSender("TEST", "***********@gmail.com");
  smtpData.setPriority("High");
  smtpData.setSubject("TEST");
  smtpData.setMessage("TEST", false);
  smtpData.addRecipient("********");
  smtpData.addAttachFile("/SCAN_34.csv");
  smtpData.addAttachFile("/SCAN_35.csv");
  Serial.print("Number of files attached: "); Serial.println(smtpData.attachFileCount());
  smtpData.setSendCallback(sendCallback);

  if (!MailClient.sendMail(smtpData))
    Serial.println("Error sending Email, " + MailClient.smtpErrorReason());
  smtpData.empty();
}
```
Flash used in 34%, Ram used in 17%
Regards Matthew.

**Comment by mobizt** [comment]
You are using the library as ESP-IDF component which run under different platform and not confirm it can run without problem.

You should debug yourself for the root cause.

**Comment by mobizt** [comment]
This may help.
https://www.esp32.com/viewtopic.php?t=9703

**Comment by mobizt** [comment]
This may help
https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/system/heap_debug.html#heap-corruption
https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/system/mem_alloc.html

**Comment by mateusz-kusmierz** [comment]
I'm using this library in arduino IDE not ESP-IDF.

**Comment by mobizt** [comment]
Update the ESP32 Arduino Core SDK (board package) to latest version.

Or increase the stack size in **app_main()** function in 

**C:\Users\YOUR_USER_NAME\AppData\Local\Arduino15\packages\esp32\hardware\esp32\1.0.1\cores\esp32\main.cpp** 

From

```C++
xTaskCreatePinnedToCore(loopTask, "loopTask", 8192, NULL, 1, &loopTaskHandle, ARDUINO_RUNNING_CORE);`
```
To
```C++
  xTaskCreatePinnedToCore(loopTask, "loopTask", 16384, NULL, 1, &loopTaskHandle, ARDUINO_RUNNING_CORE);
```

**Comment by mateusz-kusmierz** [comment]
It does not help:
```
Number of files attached: 2
Connecting to SMTP server...
Identification...
Authentication...
Sign in...
Sending Email header...
Sending Email body...
Sending attachments...
/SCAN_34.csv
/SCAN_35.csv
CORRUPT HEAP: multi_heap.c:432 detected at 0x3ffdfb40
abort() was called at PC 0x4008f067 on core 0

Backtrace: 0x4008fc5c:0x3ffaf040 0x4008fe8d:0x3ffaf060 0x4008f067:0x3ffaf080 0x4008f846:0x3ffaf0a0 0x400882bc:0x3ffaf0c0 0x400882f1:0x3ffaf0e0 0x400887b5:0x3ffaf100 0x4000beaf:0x3ffaf120 0x40153150:0x3ffaf140 0x401531df:0x3ffaf160 0x4015324d:0x3ffaf180 0x4015a9e1:0x3ffaf1a0 0x4015b064:0x3ffaf1d0 0x4014f3f2:0x3ffaf240 0x40150106:0x3ffaf260 0x4014ea79:0x3ffaf280 0x4008c145:0x3ffaf2b0

Rebooting...

```

I wonder if your code uses core0 or just only standard core1 on esp32?

**Comment by mobizt** [comment]
Are you run it in RTOS task?

Please post your full code.

I don't see how sendEmail() called from.

Did [this example](https://github.com/mobizt/ESP32-Mail-Client/blob/master/examples/Send_email/Send_email.ino) works for you?

**Comment by mateusz-kusmierz** [comment]
Sorry, might be fact that the files I tried to attach were corrupted somehow... 
Formatted SD card and now it works okay with new files.
Although files I tried to send via email opened normally in excel.

I'm not running RTOS task. sendEmail is just called by pressing physical button.

I checked every step by calling heap_caps_check_integrity_all()

And it was crashing after calling send_base64_encode_mime_file(smtpData._net->getStreamPtr(), file);

Crashed occured somewhere here:
```
    if (len - fbufIndex >= 3)
    {
      file.read(fbuf, 3);

      buf[byteAdd++] = base64_table[fbuf[0] >> 2];
      buf[byteAdd++] = base64_table[((fbuf[0] & 0x03) << 4) | (fbuf[1] >> 4)];
      buf[byteAdd++] = base64_table[((fbuf[1] & 0x0f) << 2) | (fbuf[2] >> 6)];
      buf[byteAdd++] = base64_table[fbuf[2] & 0x3f];

      dByte += 4;

```

I couldn't replicate this error with new files, so this issue is closed for now.

**Comment by mobizt** [comment]
Ok thanks for debug and let me know possible bugs.
I'll check and fix it and will inform the update.

**Comment by mobizt** [comment]
The issue was fixed, please update the library.

**Comment by mateusz-kusmierz** [comment]
Thank you. What exactly was causing the trouble?

**Comment by mobizt** [comment]
The internal index to set the base64 split chunk data is out of bounds.

To comply with rfc standard, library splits the base64 attachment data into 76 bytes of chunks which joined with CR LF bytes before sending to server.

**Comment by mateusz-kusmierz** [comment]
Thanks.
I updated this library either by arduino IDE library manager, and also by downloading zip from github, but my code does not compile now:

`C:\Users\Admin\Documents\Arduino\libraries\ESP32-Mail-Client-master\src/ESP32_MailClient.h:46:29: fatal error: ESP32TimeHelper.h: No such file or directory`

**Comment by mateusz-kusmierz** [comment]
After placing ESP32TimeHelper .cpp and .h files, inside scr dir it works.
You could add information on main page ESP32-Mail-Client, that installing ESP32-Time-Helper is now obligatory.

**Comment by mobizt** [comment]
Sorry I make some mistake in v 2.1.0 update process. 

I'm not at my computer now.
Please wait for about 1 hour when I get to my computer and push the update again.

You can download version 2.0.9 from released zip file which fix the bugs, the new version 2.1.0 includes time functions as some user requested.

**Comment by mobizt** [comment]
The time helper library used in this library has some minor modification from original one.
